### DISCUSSION TITLE
outputting mass flux diagnostics from samfdeepcnv.f using aux arrays

### DISCUSSION THREAD

**Discussion by aschuh** [ufs-community/ufs-weather-model#1862] [Q&A]
I'm struggling to output both tendencies via dtend* as well as mass flux diags ( ud_mf, dd_mf, dt_mf) from samfdeepcnv.f, using the 2D aux arrays as detailed at : https://ccpp-techdoc.readthedocs.io/en/latest/ParamSpecificOutput.html#availtendfaq . I've followed instructions and added necessary code to meta file and fortran files and updated input.nml.

In the output log file, I see that the dtend* are selected:

  0: dtend selected: gfs_phys dtend_temp_lw = temperature tendency due to long wave radiation (K s-1)
  0: dtend selected: gfs_phys dtend_temp_sw = temperature tendency due to short wave radiation (K s-1)
  0: dtend selected: gfs_phys dtend_temp_pbl = temperature tendency due to PBL (K s-1)
  0: dtend selected: gfs_phys dtend_temp_deepcnv = temperature tendency due to deep convection (K s-1)
  0: dtend selected: gfs_phys dtend_temp_shalcnv = temperature tendency due to shallow convection (K s-1)
  0: dtend selected: gfs_phys dtend_temp_mp = temperature tendency due to microphysics (K s-1)
  0: dtend selected: gfs_phys dtend_temp_orogwd = temperature tendency due to orographic gravity wave drag (K s-1)
  0: dtend selected: gfs_phys dtend_temp_cnvgwd = temperature tendency due to convective gravity wave drag (K s-1)
  0: dtend selected: gfs_phys dtend_temp_phys = temperature tendency due to physics (K s-1)
  0: dtend selected: gfs_dyn dtend_temp_nophys = temperature tendency due to non-physics processes (K s-1)
  0: dtend disabled: gfs_phys dtend_u_pbl = x wind tendency due to PBL (m s-2)

I also see that the code appears to see the aux3d adjustments:

Aug-11 14:22 ol3:/work2/noaa/co2/aschuh/stmp/aschuh/FV3_RT/rt_15823/merra2_thompson_ghg$ grep aux out
  0:   naux2d            :            0
  0:   naux3d            :            3
  0:   aux3d_time_avg    :  F F F
  0:  add field to after bilinear, fld=aux3d_01
Aug-11 14:23 ol3:/work2/noaa/co2/aschuh/stmp/aschuh/FV3_RT/rt_15823/merra2_thompson_ghg$ 

However, I'm failing to see the variables emerge in the history files.  I assume them should probably be dropped in the *phy_data.tile*nc files?  I'm using a variation on the FV3_GFS_v17_p8 suite which would seem to be supported by the web description.

input.nml:

&gfs_physics_nml
  fhzero       = 6
  h2o_phys     = .true.
  ldiag3d      = .true.
  qdiag3d      = .true.
  dtend_select = 'dtend*mp', 'dtend_temp_*'
  naux2d         = 0
  naux3d         = 3
  aux3d_time_avg = .false.,.false.,.false.
  print_diff_pgr = .false.
  fhcyc        = 24
....

Any ideas or places I should look?  switches I'm missing?

andrew

**Comment by grantfirl** [comment]
@aschuh Have you modified the diag_table per https://ccpp-techdoc.readthedocs.io/en/latest/ParamSpecificOutput.html#outputting-tendencies?

**Reply by aschuh** [reply]
Sorry, I realized I forgot to include that, *yes*, I did and it looked like:

20161003.00Z.C96.64bit.non-mono
2016 10 03 00 0 0

"fv3_history",    0,  "hours",  1,  "hours",  "time"
"fv3_history2d",  0,  "hours",  1,  "hours",  "time"

#
# Tracers
"gfs_dyn",     "co2",       "co2",     "fv3_history",  "all",  .false.,  "none",  2
"gfs_dyn",     "ch4",       "ch4",     "fv3_history",  "all",  .false.,  "none",  2
"gfs_dyn",     "sf6",       "sf6",     "fv3_history",  "all",  .false.,  "none",  2
"gfs_dyn",     "smoke",       "smoke",     "fv3_history",  "all",  .false.,  "none",  2
"gfs_phys", "dtend_qv_mp",       "dtend_qv_mp",       "fv3_history", "all", .false., "none", 2
"gfs_phys", "dtend_liq_wat_mp",  "dtend_liq_wat_mp",  "fv3_history", "all", .false., "none", 2
"gfs_phys", "dtend_rainwat_mp",  "dtend_rainwat_mp",  "fv3_history", "all", .false., "none", 2
"gfs_phys", "dtend_ice_wat_mp",  "dtend_ice_wat_mp",  "fv3_history", "all", .false., "none", 2
"gfs_phys", "dtend_snowwat_mp",  "dtend_snowwat_mp",  "fv3_history", "all", .false., "none", 2
"gfs_phys", "dtend_graupel_mp",  "dtend_graupel_mp",  "fv3_history", "all", .false., "none", 2
"gfs_phys", "dtend_cld_amt_mp",  "dtend_cld_amt_mp",  "fv3_history", "all", .false., "none", 2
"gfs_phys", "dtend_temp_phys",   "dtend_temp_phys",   "fv3_history", "all", .false., "none", 2
"gfs_dyn",  "dtend_temp_nophys", "dtend_temp_nophys", "fv3_history", "all", .false., "none", 2
#"gfs_phys",     "dtend_temp_deepcnv","dtend_temp_deepcnv", "fv3_history",  "all",  .false.,  "none",  2
#"gfs_phys",     "dtend_v_deepcnv","dtend_v_deepcnv", "fv3_history",  "all",  .false.,  "none",  2
#"gfs_phys",     "dtend_u_deepcnv","dtend_u_deepcnv", "fv3_history",  "all",  .false.,  "none",  2

# Auxiliary output
#"gfs_phys",    "aux2d_01",     "aux2d_01",      "fv3_history2d",  "all",  .false.,  "none",  2
#"gfs_phys",    "aux2d_02",     "aux2d_02",      "fv3_history2d",  "all",  .false.,  "none",  2
"gfs_phys",    "aux3d_01",     "aux3d_01",      "fv3_history",    "all",  .false.,  "none",  2

###
# FV3 variabls needed for NGGPS evaluation
###
"gfs_dyn",     "ucomp",       "ugrd",      "fv3_history",    "all",  .false.,  "none",  2
....

**Reply by aschuh** [reply]
It seems like a log message indicating that an output file is being written along w/ the variable name currently being added could be useful, but I didn't see anything like that in log file.  The only other mention of these vars is early in the log ...

  0:  creating microphysics lookup tables ...
  0:  using: mu_c_o=15.00 mu_i= 0.00 mu_r= 0.00 mu_g= 0.00
  0: Calculating Thompson tables part 1 took      0.981 seconds.
  0: Calling radar_init took      0.000 seconds.
  0: Computing rain collecting graupel table took      0.053 seconds.
  0:    creating rain collecting snow table
  0: Computing rain collecting snow table took      0.018 seconds.
  0: Computing freezing of water drops table took      0.210 seconds.
  0: Calculating Thompson tables part 2 took      0.210 seconds.
  0: NOTE from PE     0: before create fcst grid
  0:  dateS=hours since 2021-03-22 06:00:00date_init=        2021           3
  0:           22           6           0           0
  0:  in fv_phys bundle,nbdl=           2
  0:  add 3D field to after nearest_stod, fld=dtend_qv_mp
  0:  add 3D field to after nearest_stod, fld=dtend_liq_wat_mp
  0:  add 3D field to after nearest_stod, fld=dtend_ice_wat_mp
  0:  add 3D field to after nearest_stod, fld=dtend_rainwat_mp
  0:  add 3D field to after nearest_stod, fld=dtend_snowwat_mp
  0:  add 3D field to after nearest_stod, fld=dtend_graupel_mp
  0:  add 3D field to after nearest_stod, fld=dtend_temp_phys
  0:  add 3D field to after nearest_stod, fld=dtend_temp_nophys
  0:  add 3D field to after nearest_stod, fld=nwfa
  0:  add 3D field to after nearest_stod, fld=nifa
  0:  add field to after bilinear, fld=aux3d_01

**Reply by aschuh** [reply]
A bit more info here.  

When I change diag_table, relative to entries of the type "gfs_dyn" going to "fv3_history", OR relative to entries of the form "gfs_phys" going to "fv3_history2d", I get changed variables in the "atm*nc" and "sfc*nc" interpolated netcdf files.  That's good, there is a "connection".

I have added several 3d passive tracers which work when added to diag_table (via gfs_dyn) and also since they are added to field_table ( I assume this matters? ), I get them in history files on the native grid.  My desire is to get physics 3D output like aux3d, or upd_mf, which seems like it should be output when ldiag3d and qdiag3d are set to TRUE, which they are.  Preferably I'd like it both on the interpolated and native grid, but the native grid is probably more important.  However, I never see it, either in the interpolated output (would this end up in the "atm*nc" files?) or the RESTART/ files, where I presume it would end up in the phys*tile*nc files.

**Comment by aschuh** [comment]
I believe I came to a conclusion/answer on this, which can be seen here:

**Comment by aschuh** [comment]
https://github.com/ufs-community/ufs-weather-model/discussions/1867
